File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.cpp
class Solution {
public:
int sum(vector<int>& nums, int div){
    int sum=0;
    for(int i=0;i<nums.size();i++){
        sum+=ceil((double)nums[i]/(double)div);
    }
    return sum;
}
    int smallestDivisor(vector<int>& nums, int threshold) {
        int maxi=*max_element(nums.begin(),nums.end());
        int l=1,h=maxi;
        while(l<=h){
            int mid=l+(h-l)/2;
            if(sum(nums,mid)<=threshold){
                int ans=mid;
                h=mid-1;
            }
            else{
                l=mid+1;
            }

        }
        return l;
    }
};

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.java
class Solution {
    public 
    int maxi(int[] nums){
        int m=Integer.MIN_VALUE;
        for(int i=0;i<nums.length;i++){
            m=Math.max(m,nums[i]);
        }
        return m;
    }
    int sum(int[] nums,int mid){
        int sum=0;
        for(int i=0;i<nums.length;i++){
            sum+=Math.ceil((double)nums[i]/(double)mid);
        }
        return sum;
    }
    int smallestDivisor(int[] nums, int threshold) {
        int l=0;
        int h=maxi(nums);
        while(l<=h){
            int mid=l+(h-l)/2;
            if(sum(nums,mid)<=threshold)
            h=mid-1;
            else
            l=mid+1;
        }
        return l;
    }
}